### DISCUSSION TITLE
Optionally disallow add to cart on items with a products_date_available in the future

### DISCUSSION THREAD

**Discussion by scottcwilson** [zencart/zencart#5783] [Ideas]
I'd be happy to create a PR for this; just want it to be pre-approved before doing all the work. 

**Is your feature request related to a problem? Please describe.**
Some customers want to show off items which will be available in the future, but not allow them to be added to the cart. 

**Describe the solution you'd like**
Add to Cart functionality should not be presented (if so configured) on items with a future available date. 

In other words: Upcoming products could be created and enabled (so they can be displayed in the storefront), but not purchased; they would act more like Call for Price products until the Date Available was reached.

**Comment by drbyte** [comment]
Would this apply to individual products? Or would it be a product-type-specific setting? Or storewide setting?
Or would it apply only to certain attributes?

Do you envision making db schema changes? or avoiding them?

What would show in place of an add-to-cart button? What would happen if add-to-cart buttons are turned off but qty input boxes are presented instead? What about buy-now buttons?

What would happen if a customer adds it to their cart "anyway" (whether by tampering, or by restoring a saved cart from when the product was previously available, or the storeowner accidentally emails a buy-now link to everyone but then changes the product status in the db)? At what point would they be alerted to a problem, and would checkout be disallowed?

Would these show in the "new products" list? When would they show up? For how long?

Does this have any effect on categories being displayed (vs not displayed if empty) if they only contain this kind of product?

How does this differ from existing built-in functionality for when products are marked as status=0 and/or date_available is future?

Anything else?

**Comment by scottcwilson** [comment]
It would be a global setting "Do not allow add to cart on products with a date_available in the future" in Admin > Configuration > Stock.   That would be the only schema change. 

No Add to cart button.

If it gets added accidentally (example: storeowner sets date available after someone has added it to their cart) the valid to checkout logic would remove the item from the cart and issue an error message.

No other changes to display.

This differs from existing behavior because currently, a product can be added to cart even if it has a future available date.  The thinking was that the available date would trigger a change from disabled to enabled (if so configured) but no provision was made for allowing the product to be visible but not purchasable.

**Comment by mc12345678** [comment]
`Do not allow add to cart on products with a date_available in the future` is overly stated as during Out Of Box (OOB) operation the date_available is "wiped" if that date has come due. So... If trying for "descriptive" OOB installation text, perhaps `Do not allow add to cart on products with a date_available`. That would still be accurate or: `Prevent add to cart on products with a date_available` is one iteration "shorter". 

There may be other/additional ways to shorten the text, still be clear and possibly retain relatively "standard" search terms. Otherwise also thought of: `Prevent adding date_available product to cart` is a few less characters but loses the "add to cart" search result from developers toolkit.

**Comment by drbyte** [comment]
> OOB operation 
> OOB installation text

?

**Comment by drbyte** [comment]
> `date_available`

Ya, I'd stay away from mentioning anything that's db-schema-specific. Stick to phrases the storeowner understands, such as controls they see in their admin product-edit screen, like "Date Available": 
![Admin Product Date Available 2021-01-07 at 11 47 00 AM](https://user-images.githubusercontent.com/404472/103919550-25149600-50de-11eb-844b-5b12e7e768f5.jpg)

**Comment by drbyte** [comment]
> No Add to cart button.

Hmmm ... I suppose leaving it blank is an option. Something like "Coming Soon" would probably be more intuitive.

> If it gets added accidentally (example: storeowner sets date available after someone has added it to their cart) the valid to checkout logic would remove the item from the cart and issue an error message.

Interestingly, the shopping_cart class currently only checks the `allow_add_to_cart` flag on `actionBuyNow`, not elsewhere.

> This differs from existing behavior because currently, a product can be added to cart even if it has a future available date. 

Yes, as long as `products_quantity > 0` and `products_status = 1`

**Comment by drbyte** [comment]
There are some design considerations to make decisions on:

- `zen_get_products_allow_add_to_cart` is probably where one would be most tempted to add a date check.
- the shopping_cart class only references `zen_get_products_allow_add_to_cart` during `actionBuyNow`, not during other operations which add products to cart or validate their presence
- `zen_get_products_allow_add_to_cart` is used by the SNAF (I hate that acronym) modules to control whether the product's price is shown. Should price be shown in the case of "date_available is future"?

My point: perhaps `zen_get_products_allow_add_to_cart` is not the place to implement this. If it's necessary to add a date check into that function, call another function to do that, maybe `zen_get_product_allow_future_date_in_cart($product_id)` which checks your new Config Switch which then in turn calls something like `zen_is_date_future($date)` and pass the product's date_available to it and get a boolean response back.

**Comment by proseLA** [comment]
this issue makes my head spin.  i think it would be nice to re-examine our current switches, and see if there is a better way to examine a what products status means.  for example:

- enabled v disabled.  given the current schema, i have clients with disabled products, who now display those products on the storefront.  they do not come up in search, and they can not be added to the cart; but due to their SEO value, the pages are still active and return a 200.
- prior to the implementation of future dates, i implemented my own logic (and have yet to move to new ZC implementation for any number of reasons) for the use of available dates.  in this scenario, i have enabled products with future available dates whose pages do not show up on the site.  this was a way for the store owner to stage his products prior to their arrival, and not display them until they arrive in stock.  curiously enough, even though the pages do not display, the proper URL still returns a 200.

it just seems to me that it would be nice to somehow come up with a more structured, easier way to think about products status, what it means, and when a product can be added to a cart.  it is not an easy issue and goes to the heart of a shopping cart.  but i think continuing to add on to the existing schema without taking an opportunity for a really examine handling of products is going to further confuse the issue.

**Comment by torvista** [comment]
I have extra switches per product to control 
- display of price block (some "products" are for information only)
- display of Add to Cart button (hidden for above), or just hidden when out of stock. That stops it being added to the cart.
- display of Back in Stock Notification link and form
- display of stock status

I don't think product status can mean anything except active or not, while it is boolean. Making it have multiple states could encompass all of the above and allow customisation, but what a change that would be!
